### PR TITLE
refactor(tt): experiment with better page types

### PR DESCRIPTION
**Pull request by jbranchaud**
Use Next.js type helpers to get better typing across the
`getServerSideProps` and rendered component boundaries. It is not
perfect and will get more out of `tRPC` as we continue to expand how we
are using that. Nevertheless, we have a lot of pages that us
`getServerSideProps` and this can help improve our typings for those
situations.

[getServerSideProps with TypeScript](https://nextjs.org/docs/api-reference/data-fetching/get-server-side-props#getserversideprops-with-typescript)

props

### CODE REVIEW

File: apps/total-typescript/src/pages/invoices/[merchantChargeId].tsx [hunk 1]
```diff
@@ -2,24 +2,36 @@ import * as React from 'react'
...
 import {getToken} from 'next-auth/jwt'
+import {isArray} from 'lodash'
 
-export const getServerSideProps: GetServerSideProps = async ({
+type InvoiceProps = {
+  charge: Stripe.Charge
+  product: Product
+  merchantChargeId: string
+  bulkCoupon: Coupon | null
+}
```
Comment (jbranchaud) [inline apps/total-typescript/src/pages/invoices/[merchantChargeId].tsx:23]: Define the types in one place at the top of the page, then use them in `getServerSideProps` with `GetServerSideProps<InvoiceProps>`. Then the page component can infer the props it is receiving by doing `InferGetServerSidePropsType<typeof getServerSideProps>`.

File: apps/total-typescript/src/pages/invoices/[merchantChargeId].tsx [hunk 2]
```diff
@@ -46,14 +58,26 @@ export const getServerSideProps: GetServerSideProps = async ({
...
+      }
+
       return {
-        props: {
-          charge,
-          product: convertToSerializeForNextResponse(product),
-          merchantChargeId,
-          bulkCoupon: convertToSerializeForNextResponse(bulkCoupon),
-        },
+        props,
```
Comment (jbranchaud) [inline apps/total-typescript/src/pages/invoices/[merchantChargeId].tsx:80]: If I declare `props` like this (with its explicit type as `InvoiceProps`) and then return it, I get type warnings right here for the specific values that have type issues. If I assemble and return props inline without explicitly referencing the type, then the type errors show up on the function and are an indecipherable bundle.

**Review by joelhooks — COMMENTED** [review]
I don't personally like this syntax and don't think we get a lot of benefit at the cost of general readability in the code. The props were already typed in the component and we don't gain much in the way of safety in the server rendering side.

what am I missing?

**Comment by joelhooks** [comment]
https://trpc.io/docs/ssg#fetch-data-in-getstaticprops

here's this technique used in an example

**Pull request closed without merging**